tagmanager/src/tm_parser.c

```diff
@@ -0,0 +1,647 @@
+	{'g', tm_tag_enum_t},
+	{'t', tm_tag_typedef_t},
+	{'v', tm_tag_variable_t},
+	{'M', tm_tag_macro_t},
+	{'m', tm_tag_field_t},
+	{'e', tm_tag_enumerator_t},
+	{'F', tm_tag_method_t},
+};
+
+static TMParserMapEntry go_map[] = {
+	{'p', tm_tag_namespace_t},
+	{'f', tm_tag_function_t},
+	{'c', tm_tag_macro_t},
+	{'t', tm_tag_typedef_t},
+	{'v', tm_tag_variable_t},
+	{'s', tm_tag_struct_t},
+	{'i', tm_tag_interface_t},
+	{'m', tm_tag_member_t},
+};
+
+static TMParserMapEntry json_map[] = {
+	{'o', tm_tag_member_t},
+	{'a', tm_tag_member_t},
+	{'n', tm_tag_member_t},
+	{'s', tm_tag_member_t},
+	{'b', tm_tag_member_t},
+	{'z', tm_tag_member_t},
+};
+
+/* not in universal-ctags */
+static TMParserMapEntry powershell_map[] = {
+	{'f', tm_tag_function_t},
+	{'v', tm_tag_variable_t},
+};
+
+
+typedef struct
+{
+    TMParserMapEntry *entries;
+    guint size;
+} TMParserMap;
+
+#define MAP_ENTRY(map) {map, (sizeof(map)/sizeof(TMParserMapEntry))}
+
+/* keep in sync with TM_PARSER_* definitions in the header */
+static TMParserMap parser_map[] = {
+	MAP_ENTRY(c_map),
+	MAP_ENTRY(c_map),	/* C++ - same as C */
+	MAP_ENTRY(java_map),
+	MAP_ENTRY(makefile_map),
+	MAP_ENTRY(pascal_map),
+	MAP_ENTRY(perl_map),
+	MAP_ENTRY(php_map),
+	MAP_ENTRY(python_map),
+	MAP_ENTRY(latex_map),
+	MAP_ENTRY(asm_map),
+	MAP_ENTRY(conf_map),
+	MAP_ENTRY(sql_map),
+	MAP_ENTRY(docbook_map),
+	MAP_ENTRY(erlang_map),
+	MAP_ENTRY(css_map),
+	MAP_ENTRY(ruby_map),
+	MAP_ENTRY(tcl_map),
+	MAP_ENTRY(sh_map),
+	MAP_ENTRY(d_map),
+	MAP_ENTRY(f77_map),
+	MAP_ENTRY(c_map),	/* Ferite - same as C */
+	MAP_ENTRY(diff_map),
+	MAP_ENTRY(vhdl_map),
+	MAP_ENTRY(lua_map),
+	MAP_ENTRY(javascript_map),
+	MAP_ENTRY(haskell_map),
+	MAP_ENTRY(csharp_map),
+	MAP_ENTRY(freebasic_map),
+	MAP_ENTRY(haxe_map),
+	MAP_ENTRY(rest_map),
+	MAP_ENTRY(html_map),
+	MAP_ENTRY(f77_map),
+	MAP_ENTRY(c_map),	/* GLSL - same as C */
+	MAP_ENTRY(matlab_map),
+	MAP_ENTRY(vala_map),
+	MAP_ENTRY(actionscript_map),
+	MAP_ENTRY(nsis_map),
+	MAP_ENTRY(markdown_map),
+	MAP_ENTRY(txt2tags_map),
+	MAP_ENTRY(abc_map),
+	MAP_ENTRY(verilog_map),
+	MAP_ENTRY(r_map),
+	MAP_ENTRY(cobol_map),
+	MAP_ENTRY(objc_map),
+	MAP_ENTRY(asciidoc_map),
+	MAP_ENTRY(abaqus_map),
+	MAP_ENTRY(rust_map),
+	MAP_ENTRY(go_map),
+	MAP_ENTRY(json_map),
+	MAP_ENTRY(php_map),	/* Zephir - same as PHP */
+	MAP_ENTRY(powershell_map),
+};
+
+
+TMTagType tm_parser_get_tag_type(gchar kind, TMParserType lang)
+{
+	TMParserMap *map = &parser_map[lang];
+	guint i;
+
+	for (i = 0; i < map->size; i++)
+	{
+		TMParserMapEntry *entry = &map->entries[i];
+
+		if (entry->kind == kind)
+			return entry->type;
+	}
+	return tm_tag_undef_t;
+}
+
+
+gchar tm_parser_get_tag_kind(TMTagType type, TMParserType lang)
+{
+	TMParserMap *map = &parser_map[lang];
+	guint i;
+
+	for (i = 0; i < map->size; i++)
+	{
+		TMParserMapEntry *entry = &map->entries[i];
+
+		if (entry->type == type)
+			return entry->kind;
+	}
+	return '-';
+}
+
+
+void tm_parser_verify_type_mappings(void)
```

Comment:
Perhaps it would be better to split ctags parsers into two directories, those that are from upstream and those that are Geany specific.  Would make updating from upstream easier and less risky.